Add storage interface to work with multiple ways to work with data

vhugo opened the pull request:
As I suggested in #13 based on @katzien comments here is my approach to resolve the issue.

I've added the current storage type which is in memory and also added another type to store data in JSON files, to speed up the process I found github.com/nanobox-io/golang-scribble  to manage those files.

vhugo committed:
Add storage interface to work with multiple ways to save and retrieve the data

vhugo commented:
1. To create a new storage type, all you need to do is create a new file named `storage_typename.go`  with functions defined in Storage interface in `storage.go`. It would look like something similar to the following:

```go
package main

import  model "github.com/golang-bristol/beer-model"

// StorageTypeName new data storage type
type StorageTypeName struct {
	...
}

func newStorageTypeName() (*StorageTypeName, error) {
	var err error

	stg := new(StorageTypeName)

	return stg, err
}

// SaveBeer insert or update a single beer
func (s *StorageMemory) SaveBeer(beer model.Beer) error {
	var err error
	...
	return err
}

// SaveReview insert or update a single review
func (s *StorageTypeName) SaveReview(review model.Review) error {
	var err error
	...
	return err
}

// SaveBunchOfBeers insert or update a multiple beers
func (s *StorageTypeName) SaveBunchOfBeers(beers []model.Beer) error {
	var err error
	...
	return err
}

// SaveBunchOfReviews insert or update a multiple reviews
func (s *StorageTypeName) SaveBunchOfReviews(reviews []model.Review) error {
	var err error
        ...
	return err
}

// FindBeer locate full data set based on given criteria
func (s *StorageTypeName) FindBeer(criteria model.Beer) ([]*model.Beer, error) {
	var err error
	var beers []*model.Beer
	...
	return beers, err
}

// FindReview locate full data set based on given criteria
func (s *StorageTypeName) FindReview(criteria model.Review) ([]*model.Review, error) {
	var err error
	var reviews []*model.Review
	...
	return reviews, err
}
```

2. Update `storage.go` by adding a new constant for your type and add a switch case in the `newStorage` function

```go
const (
	// StorageTypeName is json type indicator
	StorageTypeName = "Name" // storage_typename.go
)

...

func newStorage(storageType string) (Storage, error) {
	var stg Storage
	var err error

	switch storageType {
	...
	case StorageTypeName:
		stg, err = newStorageTypeName()
	...
	}

	return stg, err
}
```

3. and last just update `main.go` db variable to use the new type
```go
db, err = newStorage(StorageTypeName)
if err != nil {
	log.Fatal(err)
}
```

m4-miranda commented:
Using variadic functions for SaveBeer and SaveReview could reduce redundancy. Like this:

```golang
func (s *StorageMemory) SaveBeer(beer ...model.Beer) error {
	var err error
	...
	return err
}
```

Would be called like this:
```golang
var beer1 model.Beer
var beer2 model.Beer
var manyBeers []model.Beer

db.SaveBeer(beer1)
db.SaveBeer(beer1, beer2)
db.SaveBeer(manyBeers...)
```

Fairly new to golang, would this work?

m4-miranda commented:
@katzien could you merge this PR so I can add the SQLite functions?

vhugo commented:
@m4-miranda way to go! I liked it. I'm fairly new to golang as well, so I learned something new today. 🤔
I'm going to remove "SaveBunchOf" funcs and use variadic funcs as Save.

vhugo committed:
Update save functions to work as variadic functions instead of having bulk functions

katzien commented:
Hey @m4-miranda - thanks so much for tackling this! Sorry for the delay - work got busy today so I didn't get a chance to look through and digest this properly! Will do tomorrow and merge asap :)

vhugo committed:
Merge branch 'master' of github.com:golang-bristol/beer-server into 13-persistent_storage

# Conflicts:
#	handlers.go

katzien commented on storage_mem.go:
```diff
@@ -0,0 +1,107 @@
+		}
+
+		if len(beersFound) == 1 {
+			*beersFound[0] = beer
```
Not sure why we need to do this here?

vhugo replied:
This sets a new value for an existing beer in the cellar field array. 

I thought it would be better to reuse findBeer() results to find and change current values of the struct fields.

katzien replied:
Hmm, I don't think this will work the way you think it will. FindBeer returns you a slice of pointers to beer models, but it's not tied in any way to the cellar variable. So changing the beersFound slice here will not overwrite the value in the original cellar variable. Unless that's not what you meant here?

vhugo replied:
My bad. I messed up. I'll fix this.

katzien commented on storage_mem.go:
```diff
@@ -0,0 +1,107 @@
+			*beersFound[0] = beer
+
+		} else if len(beersFound) > 1 {
+			return fmt.Errorf("can't update because found multiple entries with ID %d", beer.ID)
```
Completely up to you to decide, but I think it might be better to check for duplicate names rather than ID and auto-assign the next available ID to a beer ourselves rather than take it as a param?

vhugo replied:
This is the same point made for JSON type, the idea for `Save` function is to upsert data, so a form of identifying the record is required to update data when it matches and insert when it doesn't.

katzien commented on storage_mem.go:
```diff
@@ -0,0 +1,107 @@
+		}
+
+		if len(reviewsFound) == 1 {
+			*reviewsFound[0] = review
```
Same here as above with beers :)

vhugo replied:
Same answer for beers applies to reviews.

katzien commented on storage_mem.go:
```diff
@@ -0,0 +1,107 @@
+}
+
+// FindBeer locate full data set based on given criteria
+func (s *StorageMemory) FindBeer(criteria model.Beer) ([]*model.Beer, error) {
```
I think it will be more efficient to have separate functions for FindBeers and FindBeer. The reason being a) not having to use the blank beer as a contract for "return all" and b) FindBeers will be much more efficient if you return all data right away rather than evaluating the `if criteria == blankBeer {` every time for each beer.

katzien commented on storage_mem.go:
```diff
@@ -0,0 +1,107 @@
+}
+
+// FindReview locate full data set based on given criteria
+func (s *StorageMemory) FindReview(criteria model.Review) ([]*model.Review, error) {
```
Same here for splitting this up into FindReview and FindReviews.

katzien commented on storage.go:
```diff
@@ -0,0 +1,38 @@
+	SaveReview(...model.Review) error
+	FindBeer(model.Beer) ([]*model.Beer, error)
+	FindReview(model.Review) ([]*model.Review, error)
+}
```
Think this should have FindBeers() error and FindReviews() error as well (for efficiency / simplicity reasons explained below)

vhugo committed:
Merge branch 'master' of github.com:golang-bristol/beer-server into 13-persistent_storage

# Conflicts:
#	data.go
#	main.go

vhugo commented on storage.go:
```diff
@@ -0,0 +1,38 @@
+
+const (
+	// StorageTypeJSON is json type indicator
+	StorageTypeJSON = "JSON" // storage_json.go
```
For this scenario, I think this is a good idea. I like it. However if/when added to a configuration file this might not be easily readable. We can think about this later when the time comes.

katzien replied:
Or we can worry about the configuration file when the time comes ;)

vhugo commented on storage_json.go:
```diff
@@ -0,0 +1,154 @@
+	db *scribble.Driver
+}
+
+var (
```
To manage JSON storage I opted for a package called `golang-scribble` which organize the files in directories named after collections and filenames after resources. My concern with collection name was to be easy to identify what's what in the file system, but I agree they should be constants. 

I chose `golang-scribble` to save time and effort to implement this, I've never used it before, but I think we could improve it or even develop our own solution,  if worthwhile.

I would like to know your thoughts on that.

katzien replied:
Hmm, so golang-scribble won't work with constants?

vhugo replied:
It will. That's not what I meant, but I guess that's not important. I think I am overcomplicating things. I'll add the changes as requested. ;-)

vhugo commented on storage_json.go:
```diff
@@ -0,0 +1,154 @@
+func (s *StorageJSON) SaveBeer(beers ...model.Beer) error {
+	for _, beer := range beers {
+		var err error
+		var resource = fmt.Sprintf("%d", beer.ID)
```
As explained in the previous comment, `golang-scribble` uses resource's value as filename for each document in the directory named after the collection name, so I believe resources should work as a unique identifier for a document and could be composed as such.  

In this case, I'm only using ID, but we could combine more information that would make sense looking at the data through filesystem perspective.

And the reason to use string is that `golang-scribble` methods expect resources to be strings.

katzien replied:
Using the ID as the identifier makes sense to me. To make the intention that this needs to be a string clearer, why not use `resource := strconv.Itoa(beer.ID)`?

vhugo commented on storage_json.go:
```diff
@@ -0,0 +1,154 @@
+		}
+
+		if len(beersFound) > 0 {
+			err = s.db.Delete(collectionBeer, resource)
```
Then again,  this is more like a limitation on `golang-scribble` which has no update method, so deleting and writing was the quick option for this solution's proposal.

The idea for the `Save` function is to "upsert" data, but I think we could separate in two functions `Insert` and `Update`, so when trying to insert an already existing beer, it would return an error, and when trying to update would return an error when beer doesn't exist.

katzien replied:
That's fine if golang-scribble doesn't support update, guess we don't need to worry about it for now. I'd just make save return an error if the beer already exists rather than overwrite it - less surprises down the road that way :)

vhugo replied:
sounds good.

vhugo commented on storage_json.go:
```diff
@@ -0,0 +1,154 @@
+func (s *StorageJSON) SaveReview(reviews ...model.Review) error {
+	for _, review := range reviews {
+		var err error
+		var resource = fmt.Sprintf("%d", review.ID)
```
Same answer about collections and resources names used by `golang-scribble`, but I agree we could improve the way we assign IDs to beers and reviews.

katzien replied:
Yep, using the ID as the identifier is totally fine by me. I was just thinking of the way we assign the ID - I think it would be better to get the next available one rather than take it from the input, but I suppose that could be an improvement for later.

vhugo commented on storage_json.go:
```diff
@@ -0,0 +1,154 @@
+		var err error
+		var resource = fmt.Sprintf("%d", review.ID)
+
+		reviewsFound, err := s.FindReview(model.Review{ID: review.ID})
```
You're right. We don't. Don't even need to delete as well, the write function should overwrite the data if exists if not it will create a new document.

vhugo commented on storage_mem.go:
```diff
@@ -0,0 +1,107 @@
+
+	stg := new(StorageMemory)
+
+	return stg, err
```
You're right. The parent function can even use `new(StorageMemory)` instead.

vhugo committed:
Add few changes according to PR review

vhugo commented:
@katzien thanks for taking the time to review my PR. I'm still learning to contribute to open source projects as well as Golang. Your patience is appreciated. ;-)

Here is what my latest commit includes:

- Fix conflicts from recent merge
- Change storage type to incremental enum constants
- Fix typo in comment "action" to "actions"
- Add functions FindBeers and FindReviews to list all, removing blank check on FindBeer and FindReview functions
- Changed `fmt.Sprintf` to `strconv.Atoi` 
- Default storage for JSON type  is now ./data/
- Changed Storage JSON save function to insert only, with returning error message when duplicated record is found
- When inserting new entry the ID is based on size of the list of Beers or Reviews

Just to point what I wanted to do with the `*beersFound[0] = beer` I've fixed my mistake and also added a unit test (storage_mem_test.go) to show you what I meant.

With the recent merge containing handlers_test.go, now I can work on some tests for the storage types, but maybe it could be in a different PR, if you don't mind.

katzien commented on data.go:
```diff
@@ -79,56 +79,58 @@ func PopulateBeers() []model.Beer {
+			Name:    "XX Lager",
 			Brewery: "Cuahutemoc Moctezuma",
-			Abv: 4.2,
+			Abv:     4,
```
Ha, we should be using doubles for this not ints :D

katzien replied:
(don't feel like you need to fix this here - let's do it in a separate PR)

vhugo replied:
yeah, well. I didn't add this in the first place, but it was annoying me with error `constant 4.2 truncated to integer` when I updated the brand and was testing my code after recent merge. Anyways, I agree this shouldn't be in this PR, better create a new issue for this and new PR. I am changing it back as requested.

katzien commented on storage_json.go:
```diff
@@ -0,0 +1,173 @@
+
+		// TODO: Since delete function has not been implemented yet
+		// I think we can assume size of beers should always increase.
+		beer.ID = len(allBeers) + 1
```
This is ok to assume for now. A better way would be to make the cellar a map of ID -> beer, so we don't have to worry about ID clashes. But that's something for the future :)

katzien replied:
(Also I only thought of this now ha :D )

katzien commented on storage_mem.go:
```diff
@@ -0,0 +1,102 @@
+	for idx := range s.cellar {
+		beers = append(beers, &s.cellar[idx])
+	}
+
```
Don't think we need to do this here - we can just return s.cellar as it is :)

katzien commented on storage_mem.go:
```diff
@@ -0,0 +1,102 @@
+	for idx := range s.reviews {
+		reviews = append(reviews, &s.reviews[idx])
+	}
+
```
Same here, we should be able to return s.reviews as they are (the function signature needs to be []model.Review of course, but this is better anyway since you'd want to return a copy here, not pointers to the original slice). Same goes for FindBeers above.

vhugo committed:
Change back default data value

vhugo committed:
Change FindBeers and FindReviews return type as requested on PR review

katzien commented:
Perfect - thanks so much for all your work on this @vhugo ! And sorry for the massive delay in merging stuff in!
